From 0ca4b8577e96ba28c1af1973f52a25fdf366f57c Mon Sep 17 00:00:00 2001
From: Yaakov Selkowitz <yselkowi@redhat.com>
Date: Tue, 30 Jun 2026 16:52:43 -0400
Subject: [PATCH 2/2] feat: add ELN support to distros

ELN now identifies itself uniquely and more RHEL-like, rather than as a
variant of Fedora.

https://github.com/fedora-eln/eln/issues/565

Signed-off-by: Yaakov Selkowitz <yselkowi@redhat.com>
Signed-off-by: Jeremy Cline <jeremycline@microsoft.com>
---
 cloudinit/config/cc_ca_certs.py         |  1 +
 cloudinit/config/cc_ntp.py              | 12 ++++++++++++
 cloudinit/config/cc_resolv_conf.py      |  1 +
 cloudinit/config/cc_yum_add_repo.py     |  1 +
 cloudinit/distros/__init__.py           |  1 +
 cloudinit/distros/eln.py                |  7 +++++++
 cloudinit/net/sysconfig.py              |  1 +
 cloudinit/sources/DataSourceRbxCloud.py |  2 +-
 cloudinit/util.py                       |  1 +
 config/cloud.cfg.tmpl                   |  4 ++--
 systemd/cloud-init-generator.tmpl       |  2 +-
 systemd/cloud-init-local.service.tmpl   |  2 +-
 systemd/cloud-init-main.service.tmpl    |  2 +-
 systemd/cloud-init-network.service.tmpl |  2 +-
 14 files changed, 32 insertions(+), 7 deletions(-)
 create mode 100644 cloudinit/distros/eln.py

diff --git a/cloudinit/config/cc_ca_certs.py b/cloudinit/config/cc_ca_certs.py
index 939b62d07..9266d8848 100644
--- a/cloudinit/config/cc_ca_certs.py
+++ b/cloudinit/config/cc_ca_certs.py
@@ -72,6 +72,7 @@ distros = list(DISTRO_FAMILY.keys()) + [
     "aosc",
     "alpine",
     "debian",
+    "eln",
     "raspberry-pi-os",
     "rhel",
     "opensuse",
diff --git a/cloudinit/config/cc_ntp.py b/cloudinit/config/cc_ntp.py
index e1a27b8bc..37536dc3b 100644
--- a/cloudinit/config/cc_ntp.py
+++ b/cloudinit/config/cc_ntp.py
@@ -31,6 +31,7 @@ distros = [
     "cloudlinux",
     "cos",
     "debian",
+    "eln",
     "eurolinux",
     "fedora",
     "freebsd",
@@ -146,6 +147,14 @@ DISTRO_CLIENT_CONFIG: Dict[str, Dict] = {
             "confpath": "/etc/chrony/chrony.conf",
         },
     },
+    "eln": {
+        "ntp": {
+            "service_name": "ntpd",
+        },
+        "chrony": {
+            "service_name": "chronyd",
+        },
+    },
     "freebsd": {
         "ntp": {
             "confpath": "/etc/ntp.conf",
@@ -375,6 +384,9 @@ def generate_server_names(distro):
         # For legal reasons x.pool.sles.ntp.org does not exist,
         # use the opensuse pool
         pool_distro = "opensuse"
+    elif distro == "eln":
+        # ELN is from the Fedora project
+        pool_distro = "fedora"
     elif distro == "alpine" or distro == "eurolinux":
         # Alpine-specific pool (i.e. x.alpine.pool.ntp.org) does not exist
         # so use general x.pool.ntp.org instead. The same applies to EuroLinux
diff --git a/cloudinit/config/cc_resolv_conf.py b/cloudinit/config/cc_resolv_conf.py
index 920c539d9..8a76e4172 100644
--- a/cloudinit/config/cc_resolv_conf.py
+++ b/cloudinit/config/cc_resolv_conf.py
@@ -28,6 +28,7 @@ meta: MetaSchema = {
     "distros": [
         "alpine",
         "azurelinux",
+        "eln",
         "fedora",
         "mariner",
         "opensuse",
diff --git a/cloudinit/config/cc_yum_add_repo.py b/cloudinit/config/cc_yum_add_repo.py
index 66e718015..95974d223 100644
--- a/cloudinit/config/cc_yum_add_repo.py
+++ b/cloudinit/config/cc_yum_add_repo.py
@@ -25,6 +25,7 @@ meta: MetaSchema = {
         "azurelinux",
         "centos",
         "cloudlinux",
+        "eln",
         "eurolinux",
         "fedora",
         "mariner",
diff --git a/cloudinit/distros/__init__.py b/cloudinit/distros/__init__.py
index 2703ebd83..b7c383514 100644
--- a/cloudinit/distros/__init__.py
+++ b/cloudinit/distros/__init__.py
@@ -76,6 +76,7 @@ OSFAMILIES = {
         "azurelinux",
         "centos",
         "cloudlinux",
+        "eln",
         "eurolinux",
         "fedora",
         "mariner",
diff --git a/cloudinit/distros/eln.py b/cloudinit/distros/eln.py
new file mode 100644
index 000000000..6ceaf6d8b
--- /dev/null
+++ b/cloudinit/distros/eln.py
@@ -0,0 +1,7 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+
+from cloudinit.distros import rhel
+
+
+class Distro(rhel.Distro):
+    pass
diff --git a/cloudinit/net/sysconfig.py b/cloudinit/net/sysconfig.py
index 5d2e2e1b3..1382403d1 100644
--- a/cloudinit/net/sysconfig.py
+++ b/cloudinit/net/sysconfig.py
@@ -24,6 +24,7 @@ KNOWN_DISTROS = [
     "almalinux",
     "centos",
     "cloudlinux",
+    "eln",
     "eurolinux",
     "fedora",
     "miraclelinux",
diff --git a/cloudinit/sources/DataSourceRbxCloud.py b/cloudinit/sources/DataSourceRbxCloud.py
index 2fba1149d..dc21aa427 100644
--- a/cloudinit/sources/DataSourceRbxCloud.py
+++ b/cloudinit/sources/DataSourceRbxCloud.py
@@ -60,7 +60,7 @@ def _sub_arp(cmd):
 
 def gratuitous_arp(items, distro):
     source_param = "-S"
-    if distro.name in ["fedora", "centos", "rhel"]:
+    if distro.name in ["fedora", "eln", "centos", "rhel"]:
         source_param = "-s"
     for item in items:
         try:
diff --git a/cloudinit/util.py b/cloudinit/util.py
index 5fa0f94c7..e2eed8536 100644
--- a/cloudinit/util.py
+++ b/cloudinit/util.py
@@ -607,6 +607,7 @@ def _get_variant(info):
             "centos",
             "cloudlinux",
             "debian",
+            "eln",
             "eurolinux",
             "fedora",
             "mariner",
diff --git a/config/cloud.cfg.tmpl b/config/cloud.cfg.tmpl
index 3926ea8b4..fa4f72833 100644
--- a/config/cloud.cfg.tmpl
+++ b/config/cloud.cfg.tmpl
@@ -2,7 +2,7 @@
 # The top level settings are used as module
 # and base configuration.
 {% set is_bsd = variant in ["dragonfly", "freebsd", "netbsd", "openbsd"] %}
-{% set is_rhel = variant in ["almalinux", "centos", "cloudlinux", "eurolinux",
+{% set is_rhel = variant in ["almalinux", "centos", "cloudlinux", "eln", "eurolinux",
                              "miraclelinux", "rhel", "rocky", "virtuozzo"] %}
 {% set gecos = ({"amazon": "EC2 Default User", "azurelinux": "Azure Linux",
                  "centos": "Cloud User", "debian": "Debian",
@@ -26,7 +26,7 @@
                   "openbsd": "/bin/ksh"}) %}
 {% set usernames = ({"amazon": "ec2-user", "azurelinux": "azureuser", "centos": "cloud-user",
                      "openmandriva": "omv", "raspberry-pi-os": "pi",
-                     "rhel": "cloud-user",
+                     "eln": "cloud-user", "rhel": "cloud-user",
                      "unknown": "ubuntu"}) %}
 {% if is_bsd %}
 syslog_fix_perms: root:wheel
diff --git a/systemd/cloud-init-generator.tmpl b/systemd/cloud-init-generator.tmpl
index 888bad535..24324f204 100644
--- a/systemd/cloud-init-generator.tmpl
+++ b/systemd/cloud-init-generator.tmpl
@@ -20,7 +20,7 @@ CLOUD_SYSTEM_TARGET="/usr/lib/systemd/system/cloud-init.target"
 {% else %}
 CLOUD_SYSTEM_TARGET="/lib/systemd/system/cloud-init.target"
 {% endif %}
-{% if variant in ["almalinux", "azurelinux", "centos", "cloudlinux", "eurolinux", "fedora",
+{% if variant in ["almalinux", "azurelinux", "centos", "cloudlinux", "eln", "eurolinux", "fedora",
                   "miraclelinux", "openeuler", "OpenCloudOS", "openmandriva", "rhel", "rocky", "TencentOS", "virtuozzo"] %}
     dsidentify="/usr/libexec/cloud-init/ds-identify"
 {% elif variant == "benchmark" %}
diff --git a/systemd/cloud-init-local.service.tmpl b/systemd/cloud-init-local.service.tmpl
index a417f09ae..1e3170944 100644
--- a/systemd/cloud-init-local.service.tmpl
+++ b/systemd/cloud-init-local.service.tmpl
@@ -9,7 +9,7 @@ Wants=network-pre.target
 After=hv_kvp_daemon.service
 Before=network-pre.target
 Before=shutdown.target
-{% if variant in ["almalinux", "azurelinux", "cloudlinux", "rhel"] %}
+{% if variant in ["almalinux", "azurelinux", "cloudlinux", "eln", "rhel"] %}
 Before=firewalld.target
 {% endif %}
 {% if variant in ["ubuntu", "unknown", "debian", "raspberry-pi-os"] %}
diff --git a/systemd/cloud-init-main.service.tmpl b/systemd/cloud-init-main.service.tmpl
index 5bb27586d..d23a8fe2d 100644
--- a/systemd/cloud-init-main.service.tmpl
+++ b/systemd/cloud-init-main.service.tmpl
@@ -11,7 +11,7 @@ Description=Cloud-init: Single Process
 {% if variant in ["almalinux", "azurelinux", "cloudlinux", "ubuntu", "unknown", "debian", "raspberry-pi-os", "rhel", "fedora"] %}
 DefaultDependencies=no
 {% endif %}
-{% if variant in ["almalinux", "azurelinux", "cloudlinux", "rhel"] %}
+{% if variant in ["almalinux", "azurelinux", "cloudlinux", "eln", "rhel"] %}
 Requires=dbus.socket
 After=dbus.socket
 {% endif %}
diff --git a/systemd/cloud-init-network.service.tmpl b/systemd/cloud-init-network.service.tmpl
index c1a7c3f64..c08d8e1ec 100644
--- a/systemd/cloud-init-network.service.tmpl
+++ b/systemd/cloud-init-network.service.tmpl
@@ -15,7 +15,7 @@ After=systemd-networkd-wait-online.service
 {% if variant in ["ubuntu", "unknown", "debian", "raspberry-pi-os"] %}
 After=networking.service
 {% endif %}
-{% if variant in ["almalinux", "azurelinux", "centos", "cloudlinux", "eurolinux", "fedora",
+{% if variant in ["almalinux", "azurelinux", "centos", "cloudlinux", "eln", "eurolinux", "fedora",
                   "miraclelinux", "openeuler", "OpenCloudOS", "openmandriva",
                   "rhel", "rocky", "suse", "TencentOS", "virtuozzo"] %}
 After=NetworkManager.service
-- 
2.55.0

